### DISCUSSION TITLE
[FR] Merge playback devices and dynamically choose best protocol to use

### DISCUSSION THREAD

**Discussion by fonix232** [music-assistant/support#1518] [Feature requests and ideas]
In the current implementation, every single protocol's scanner returns a distinct playback device entity. This can result in multiple entries for the same device over different protocols, which can be quite annoying in a mixed-protocol environment.

For example, I have a number of Sonos speakers, SOME of them with AirPlay2 support, a number of Apple HomePods, a Samsung soundbar that supports both DLNA and AirPlay2, which is connected to a Philips TV that also supports DLNA, with Chromecast...

The result is a quite noisy list of devices where it's hard to pick which one I want to play on. 

For this reason, I'd like to recommend a merge logic for these speakers.

The first step would be merging ALL devices under the same IP address. If a Sonos speaker responds to both Sonos and DLNA and AirPlay2, don't create three entries, but rather one entry, with a list of protocols supported, and then when playback is initiated, use the most fitting backend.

An additional step would be filtering devices based on name, and if a non-unique name is found, attaching the device model/manufacturer behind it. In my case, both my Apple TV (which has two HomePods in Home Theatre mode connected), and my Symfonisk Sonos speaker are called "Living Room" - making it hard to figure out which is which. With this recommendation, they'd become "Living Room (Apple TV)" and "Living Room (Sonos)", making the distinction clear.

This would obviously add an extra layer between the playback engine and the devices (a sort of device aggregator, where the PBE decides which device(s) it is going to play on, and the aggregator picks the best protocol available for the selection, and routes the stream), but it would also simplify the mixed-protocol setups many people would have.

**Comment by marcelveldt** [comment]
While I understand this FR, it is in practice impossible to implement because we can not foresee which protocol is preferred by the user.
Current implementation is to enable the common/default/primary protocol by default and disable others.

For example for Sonos speakers we disable the airplay and dlna variant by default.

The idea is nice to create this intermediate layer in between but it will make the code and UI an awful lot more complex if its even possible to detect that devices should be merged.

What we can do better is the whole detection mechanism of players and presentation.
Maybe we should detect all players by default disabled and then with quick toggles to enable a player based on the protocol you want to use. In a mixed environment you see often that Airplay is used as streaming protocol for example

**Reply by fonix232** [reply]
IMO the whole of the protocol choice should be hidden from the users - especially if MA wants to target more consumer-level users, and not prosumers/developers like ourselves. Regular users don't give a damn about protocols etc., all they know is that they want the following flow:

```
Select music on controller -> [black box mechanism of MA] -> Sound comes out of selected speakers
```

I understand the need for being able to select the specific protocol - although aside from testing, I don't see the benefit over choosing just the device and letting MA select the best protocol encompassing all selected devices. It could be an optional selector in settings, maybe call it "simplified device selection" (although obviously the backend code would be a bit dirtier).

> The idea is nice to create this intermediate layer in between but it will make the code and UI an awful lot more complex if its even possible to detect that devices should be merged.

The code, maybe, but the UI not necessarily - the UI should already be feeding from a generic device list provider, which collects all devices from every enabled protocol, and lists them. Creating a wrapper layer between the enabled protocols' device list providers, and the web UI's device list displayer should be straightforward if the underlying design is solid. I'll have to check the code, though, when I have some free time.

> Current implementation is to enable the common/default/primary protocol by default and disable others.

See this is my issue here. I have a Sonos speaker and two HomePods in my living room, and I can't group them in MA, because MA picks the Sonos protocol over AirPlay.

But I also have _another_ Sonos speaker, which does not support AirPlay (a dumb old Play:1 which I'll be probably replacing anyway, but that's not the point), and sometimes I want to pair the living room Sonos with this one, in which case it should use the Sonos protocol. 

In the current approach, to achieve this, I'd need to fumble around to be able to group these devices, disabling protocols, etc., which isn't optimal.

On the other hand if we had this unifying layer that primarily detects _devices_, and each device has a list of _protocols_ it supports (mDNS can be resolved to IP and vice versa, so unifying them based on IP should be straightforward. The only outlier I can see here is if there's some funky double NAT-ting going on on someone's network, but that's already a very wonky edge case to support). Then when grouped/linked for playback, find the common denominator, and use that protocol. That way if e.g. I create a "Living Room" group with the HomePods and the AP2 capable Sonos, it uses AirPlay2, whereas if I create a "Just Sonos" group with the two Sonos speakers, it'll use the Sonos protocol.

This would make the UI less cluttered, as well as less reliant on the user for selecting the right protocols.

**Reply by marcelveldt** [reply]
The idea is intriguing but to realize it would be really difficult. Let's leave this FR open to look at once we have released 2.0 stable and explore options. What makes it complicated is that you can also have native sync/group support where you need to figure out dynamically what player type is intended to be grouped etc.

Imo the idea of consolidating the different protocols of a device is great but we should not hide the protocol from the user. So let the user choose which protocol is intended to be used (and propose a default), just like the Spotify app, which already combined speakers that have both cast and Spotify connect support.
That way you can satisfy the non-technical users (by auto picking the recommended protocol) but also the advanced use cases where, for example, a sonos speaker is targeted with airplay because it needs to play music in sync with other airplay capable speakers.

Like I said, I like the idea but at this time there is too much other work to do before we could start looking at this but once 2.0 is released I'd definitely like to talk a bit more on this subject. Thanks for starting the discussion

**Reply by OzGav** [reply]
Has 2.0 and the work on the AUDIO PIPELINE overtaken this request?

**Comment by OzGav** [comment]
This has been implemented in MA 2.8

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
